pallets/gated-marketplace/src/functions.rs

```diff
@@ -830,6 +843,84 @@ impl<T: Config> Pallet<T> {
 		Ok(())
 	}
 
+	pub fn do_block_user(
+		authority: T::AccountId,
+		marketplace_id: [u8; 32],
+		user: T::AccountId,
+	) -> DispatchResult {
+		// ensure the marketplace exists
+		ensure!(<Marketplaces<T>>::contains_key(marketplace_id), Error::<T>::MarketplaceNotFound);
+		// ensure the origin is owner or admin
+		Self::is_authorized(authority.clone(), &marketplace_id, Permission::EnlistBlockedUser)?;
+		//ensure the user is not already a participant of the marketplace
+		ensure!(
+			Self::is_authorized(
+				user.clone(), 
+				&marketplace_id,
+				 Permission::EnlistBuyOffer
+				).is_err(), 
+				Error::<T>::UserAlreadyParticipant
+			);
+		// check if the user is already blocked
+		if Self::try_unblock_user(user.clone(), marketplace_id).is_err() {
+			// if the user is not blocked, block it
+			<BlockedUsersByMarketplace<T>>::try_mutate(marketplace_id, |blocked_list| blocked_list.try_push(user.clone()))
+				.map_err(|_| Error::<T>::ExceedMaxBlockedUsers)?;
+			Self::deposit_event(Event::UserBlocked(marketplace_id, user.clone()));
+			if let Ok(application_id) = <ApplicationsByAccount<T>>::try_get(user.clone(), marketplace_id){
```

Comment:
For now, let's not remove any information from the participant (application, offers, ...). It'll work as a record. We could talk about that after the refactoring.